src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -133,13 +133,59 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
 
     private fun checkUseSpeck(holder: RsAnnotationHolder, useSpeck: RsUseSpeck) {
         if (useSpeck.isStarImport || useSpeck.useGroup != null) return
+
+        checkDuplicateImport(holder, useSpeck)
+        checkReexports(holder, useSpeck)
+    }
+
+    private fun checkDuplicateImport(holder: RsAnnotationHolder, useSpeck: RsUseSpeck) {
         val duplicates = holder.currentAnnotationSession.duplicatesByNamespace(useSpeck.containingMod, false)
         if (useSpeck.namespaces.any { useSpeck in duplicates[it].orEmpty() }) {
             val identifier = PsiTreeUtil.getDeepestLast(useSpeck)
             RsDiagnostic.DuplicateImportError(identifier).addToHolder(holder)
         }
     }
 
+    private fun checkReexports(holder: RsAnnotationHolder, useSpeck: RsUseSpeck) {
+        val item = useSpeck.ancestorStrict<RsUseItem>() ?: return
+        if (item.visibility == RsVisibility.Private) return
+
+        val path = useSpeck.path ?: return
+        val targets = path.reference?.multiResolve()?.filterIsInstance<RsItemElement>() ?: emptyList()
+
+        // targets not accessible from the use element will be handled by E0603
+        val visibleTargets = targets.filter { it.isVisibleFrom(useSpeck.containingMod) }
+        val invalidTargets = visibleTargets.filter {
+            !canBeReexported(item.visibility, it.visibility)
+        }
+
+        // if at least one item can be re-exported, allow the reexport
+        if (invalidTargets.size < targets.size) return
```

Comment:
nit: maybe it's worth to add `|| invalidTargets.isEmpty()` condition to avoid `firstOrNull()?.let` below?